source/iterator/snapshot_iterator.go

```diff
@@ -0,0 +1,401 @@
+
+	return nil
+}
+
+// retries the function until it returns false or an error
+type RetryFunc func() (bool, error)
+
+// retries supplied function using retry backoff strategy.
+func (s *SnapshotIterator) withRetry(r RetryFunc) error {
+	b := &backoff.Backoff{
+		Max:    2 * time.Minute,
+		Min:    10 * time.Second,
+		Factor: 1.1,
+		// Jitter: true,
+	}
+	for {
+		retry, err := r()
+		if err != nil {
+			return err
+		}
+		if retry {
+			d := b.Duration()
+			time.Sleep(b.Duration())
+			if d == b.Max {
+				b.Reset()
+			}
+			continue
+		} else if !retry {
+			break
+		}
+	}
+	return nil
+}
+
+// prepares and returns record in sdk.Record format. If process fails for any reason, it returns error.
+func (s *SnapshotIterator) prepareRecord(ctx context.Context, data []string) (sdk.Record, error) {
+	logger := sdk.Logger(ctx).With().Str("Method", "prepareRecord").Logger()
+	logger.Trace().Msg("Starting the prepareRecord method")
+	var dataMap = marketoclient.GetDataMap(s.fields, data)
+	createdAt, err := time.Parse(time.RFC3339, fmt.Sprintf("%s", dataMap["createdAt"]))
+	if err != nil {
+		logger.Err(err).Msg("Error while parsing createdAt")
+		return sdk.Record{}, fmt.Errorf("error parsing createdAt %w", err)
+	}
+	updatedAt, err := time.Parse(time.RFC3339, fmt.Sprintf("%s", dataMap["updatedAt"]))
+	if err != nil {
+		logger.Err(err).Msg("Error while parsing updatedAt")
+		return sdk.Record{}, fmt.Errorf("error parsing updatedAt %w", err)
+	}
+	if updatedAt.After(s.lastMaxModied) {
+		s.lastMaxModied = updatedAt
+	}
+	position := position.Position{
+		Key:       (dataMap["id"].(string)),
+		CreatedAt: createdAt,
+		UpdatedAt: updatedAt,
+		Type:      position.TypeSnapshot,
+	}
+	pos, err := position.ToRecordPosition()
+	if err != nil {
+		logger.Err(err).Msg("Error while converting position to record position")
+		return sdk.Record{}, fmt.Errorf("error converting position to record position %w", err)
+	}
+	rec := sdk.Record{
+		Payload: sdk.StructuredData(dataMap),
+		Metadata: map[string]string{
+			"id":        position.Key,
+			"createdAt": createdAt.Format(time.RFC3339),
+			"updatedAt": updatedAt.Format(time.RFC3339),
+		},
+		Position: pos,
+		Key: sdk.StructuredData{
+			"id": position.Key,
+		},
+	}
+
+	return rec, nil
+}
+
+// returns Last date from the supplied position.if p is zero value, then it queries least date from the database.
+func (s *SnapshotIterator) getLastProcessedDate(ctx context.Context, p position.Position) (time.Time, error) {
+	logger := sdk.Logger(ctx).With().Str("Method", "getInitialDate").Logger()
+	logger.Trace().Msg("Starting the getInitialDate method")
+	var date = p.CreatedAt.Add(1 * time.Second)
+	var err error
+	if reflect.ValueOf(p).IsZero() {
+		date, err = s.getLeastDate(ctx, *s.client)
+		if err != nil {
+			sdk.Logger(ctx).Error().Err(err).Msg("Failed to get the oldest date from marketo")
+			return time.Time{}, err
+		}
+	}
+	return date, nil
+}
+
+// return least date from marketo.
```

Comment:
Least date of what? All leads?

Reply:
Yes, so we can start from that date to time.Now() for snapshot iterator